ath79: add support for Compex WPJ558

hamtinepeng opened the pull request:
Specifications:
- SoC: QCA9558
- DRAM: 128MB DDR2
- Flash: 16MB SPI-NOR
- Wireless: on-board abgn 2×2 2.4GHz radio
- Ethernet: 2x 10/100/1000 Mbps (1x 802.11af PoE)
- miniPCIe slot

Flash instruction:
- From u-boot

tftpboot 0x80500000 openwrt-ath79-generic-compex_wpj558-16m-squashfs-sysupgrade.bin
erase 0x9f030000 +$filesize
cp.b $fileaddr 0x9f030000 $filesize
boot

- From cpximg loader

The cpximg loader can be started either by holding the reset button
during power up. Once it's running, a TFTP-server under 192.168.1.1 will accept
the image appropriate for the board revision that is etched on the board.

For example, if the board is labelled '6A07':

tftp -v -m binary 192.168.1.1 -c put openwrt-ath79-generic-compex_wpj558-16m-squashfs-cpximg-6a07.bin

Signed-off-by: Romain Mahoux <romain@mahoux.fr>

CodeFetch commented on target/linux/ath79/dts/qca9558_compex_wpj558-16m.dts:
```diff
@@ -0,0 +1,151 @@
+			label = "green:sig3";
+			gpios = <&gpio 22 GPIO_ACTIVE_LOW>;
+		};
+
```
What are these LEDs typically used for?

CodeFetch replied:
Do you mean link indicators for the ethernet ports? Because that would make much more sense to me. Link indicators are very useful. Please use the LEDs for that.

adschm replied:
@CodeFetch NO. Please don't just give random "advice" here based on what you might have liked.

LEDs should generally follow the behavior of OEM firmware _if_ possible and reasonable. Otherwise they are left undefined _by default_.

hamtinepeng replied:
The hardware guide call them "RSSI" LEDs, it means that depending on the wireless signal strength, some of the four LEDs light up. However I have just flashed the stock firmware (WPJ558-v6.2.12_b190718.img) and the four LEDs are turned off by default. The user has to configure them manually depending on its wishes via the Compex web interface.
@CodeFetch You see the equipment as a finished product whereas it is a reference/hacker board.
IMHO we must let them disabled like default OEM firmware behavior.

adschm replied:
Based on the information I got here, I'm fine with the current LED setup.

CodeFetch commented on target/linux/ath79/dts/qca9558_compex_wpj558-16m.dts:
```diff
@@ -0,0 +1,151 @@
+		led_sig4: sig4 {
+			label = "green:sig4";
+			gpios = <&gpio 23 GPIO_ACTIVE_LOW>;
+		};
```
Maybe call it status then like other boards... If the other LEDs have no other use, it could make sense to use the green LEDs as port indicators, the red one for upgrade and failsafe and the yellow one as running and boot or something...

adschm replied:
If they are originally signal strength indicators, using sig1 to sig4 is essentially fine. One should name them after their original purpose, and not rename just because we hijack them as system leds. So, based on that the current scheme is basically fine.

One optional improvement could be to include the strength into the name as some other device do, e.g. use green:signal-low, green:signal-mediumlow, green:signal-mediumhigh, green:signal-high or just green:low, green:mediumlow etc. But whether that's really better depends on certains aspects, and taste.

adschm replied:
Addendum: Since this actually an ar71xx device, I'd keep the sigX names in this case, as they are consistent with the old names and there is no strong reason to change.

hamtinepeng replied:
Ok il will finally keep the sigX name and replace led_sig4 by status

adschm replied:
No, please also keep led_sig4. This should be named after the led actually used, so from reading the lines in aliases node one can immediately understand which led is used.

hamtinepeng replied:
Ok thank you for the explanation

adschm commented on target/linux/ath79/dts/qca9558_compex_wpj558-16m.dts:
```diff
@@ -0,0 +1,151 @@
+&spi {
+	status = "okay";
+
+	num-cs = <1>;
```
num-cs should be dropped:
https://github.com/openwrt/openwrt/commit/72bd92bea0aacf9a008782bea9e015c3ac0a2888

adschm commented on target/linux/ath79/dts/qca9558_compex_wpj558-16m.dts:
```diff
@@ -0,0 +1,151 @@
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <33000000>;
+		m25p,fast-read;
```
Is this really necessary? Typically, fast-read is only required above 50 MHz.

adschm replied:
I don't understand what you mean? 33 MHz is only available for fast-read, and normal read should be lower?

hamtinepeng replied:
I double check the mx25l12805d datasheet and indeed the fast access time is at 50 MHz. Normal read is between 10KHz & 33 MHz.

# flash_speed /dev/mtd3 # 33 MHz               | # flash_speed /dev/mtd3 # 50 MHz
not NAND flash, assume page size is 512 bytes. | not NAND flash, assume page size is 512 bytes.
scanning for bad eraseblocks                   | scanning for bad eraseblocks
scanned 1 eraseblocks, 0 are bad               | scanned 1 eraseblocks, 0 are bad
testing eraseblock read speed                  | testing eraseblock read speed
eraseblock read speed is 438 KiB/s             | eraseblock read speed is 4000 KiB/s
testing page read speed                        | testing page read speed
page read speed is 435 KiB/s                   | page read speed is 3764 KiB/s
testing 2 page read speed                      | testing 2 page read speed
2 page read speed is 744 KiB/s                 | 2 page read speed is 4000 KiB/s
Testing 2x multi-block erase speed             | Testing 2x multi-block erase speed
2x multi-block erase speed is 224 KiB/s        | 2x multi-block erase speed is 228 KiB/s
Testing 4x multi-block erase speed             | Testing 4x multi-block erase speed
4x multi-block erase speed is 219 KiB/s        | 4x multi-block erase speed is 230 KiB/s
Testing 8x multi-block erase speed             | Testing 8x multi-block erase speed
8x multi-block erase speed is 223 KiB/s        | 8x multi-block erase speed is 227 KiB/s
Testing 16x multi-block erase speed            | Testing 16x multi-block erase speed
16x multi-block erase speed is 226 KiB/s       | 16x multi-block erase speed is 229 KiB/s
Testing 32x multi-block erase speed            | Testing 32x multi-block erase speed
32x multi-block erase speed is 220 KiB/s       | 32x multi-block erase speed is 228 KiB/s
Testing 64x multi-block erase speed            | Testing 64x multi-block erase speed
64x multi-block erase speed is 225 KiB/s       | 64x multi-block erase speed is 229 KiB/s
finished                                       | finished

adschm commented on target/linux/ath79/dts/qca9558_compex_wpj558-16m.dts:
```diff
@@ -0,0 +1,151 @@
+	status = "okay";
+
+	mtd-cal-data = <&art 0x1000>;
+	mtd-mac-address = <&art 0x10>;
```
That's a very uncommon location. Are you sure?

adschm commented on target/linux/ath79/dts/qca9558_compex_wpj558-16m.dts:
```diff
@@ -0,0 +1,151 @@
+
+	pll-data = <0x03000101 0x00000101 0x00001616>;
+	mtd-mac-address = <&uboot 0x2e010>;
+	mtd-mac-address-increment = <2>;
```
Nothing in 0x2e016?

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/02_network:
```diff
@@ -202,6 +202,9 @@ ath79_setup_interfaces()
 			"0@eth0" "3:lan" "2:wan"
 		;;
+	compex,wpj558-16m)
+		ucidef_add_switch "switch0" "1:lan" "5:wan" "6@eth0"
```
Please use the same formatting as everybody else.

adschm replied:
I do not understand the question? Test which ports are connected to the switch and then assign them so LAN and WAN is at the same ports as with stock firmware. PoE should not be relevant at all here?

adschm replied:
ar71xx was using opposite assignment: 5:lan 1:wan ...

adschm replied:
You only use one eth here although you set up two in DTS. This needs further work ...

hamtinepeng replied:
eth1 must be wired to port 0 because it is the last available port for RGMII on AR8327

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/02_network:
```diff
@@ -488,6 +491,7 @@ ath79_setup_macs()
 		wan_mac=$(fritz_tffs -n macb -i $(find_mtd_part "tffs (1)"))
 		;;
 	compex,wpj344-16m|\
+	compex,wpj558-16m|\
```
You set up two ethX in DTS, so please set the address there.

Since you only use one of these in 02_network, though, it's probably just your configuration being wrong and this may be kept eventually.

adschm commented on target/linux/ath79/image/generic.mk:
```diff
@@ -642,6 +642,19 @@ define Device/compex_wpj531-16m
+  DEVICE_VARIANT := 16M
+  SUPPORTED_DEVICES += wpj558
+  IMAGES += cpximg-6a07.bin
+  IMAGE/cpximg-6a07.bin := append-kernel | pad-to $$$$(BLOCKSIZE) | append-rootfs | pad-rootfs | mkmylofw_16m 0x688 1
```
Please wrap lines after 74 characters max.

adschm commented on target/linux/ath79/image/generic.mk:
```diff
@@ -642,6 +642,19 @@ define Device/compex_wpj531-16m
+  DEVICE_VENDOR := Compex
+  DEVICE_MODEL := WPJ558
+  DEVICE_VARIANT := 16M
+  SUPPORTED_DEVICES += wpj558
```
Is the vendor firmware OpenWrt or why do you need this?

adschm replied:
Ah, so this was supported with ar71xx. Somebody could have told me earlier ...
In this case, SUPPORTED_DEVICES is correct.

adschm commented:
Your DTS setup does not match your setup in 02_network...

adschm commented on target/linux/ath79/dts/qca9558_compex_wpj558-16m.dts:
```diff
@@ -0,0 +1,151 @@
+
+	wifi@0,0 {
+		reg = <0x0000 0 0 0 0>;
+	};
```
You do not tell anything about a second wifi in the commit message?

hamtinepeng replied:
Wifi on miniPCIe is optional but I will five more details on the commit message because of the wifi node.

adschm replied:
if it's optional, please don't include it _by default_.

hamtinepeng replied:
Do I need to remove the whole wifi node?

adschm replied:
Well, if wifi is not present by default, there is no need to have a wifi node here.

hamtinepeng committed:
ath79: add support for Compex WPJ558

Specifications:
- SoC: QCA9558
- DRAM: 128MB DDR2
- Flash: 16MB SPI-NOR
- Wireless: on-board abgn 2×2 2.4GHz radio
- Ethernet: 2x 10/100/1000 Mbps (1x 802.11af PoE)
- miniPCIe slot

Flash instruction:
- From u-boot

tftpboot 0x80500000 openwrt-ath79-generic-compex_wpj558-16m-squashfs-sysupgrade.bin
erase 0x9f030000 +$filesize
cp.b $fileaddr 0x9f030000 $filesize
boot

- From cpximg loader

The cpximg loader can be started either by holding the reset button
during power up. Once it's running, a TFTP-server under 192.168.1.1 will accept
the image appropriate for the board revision that is etched on the board.

For example, if the board is labelled '6A07':

tftp -v -m binary 192.168.1.1 -c put openwrt-ath79-generic-compex_wpj558-16m-squashfs-cpximg-6a07.bin

Signed-off-by: Romain Mahoux <romain@mahoux.fr>

adschm commented on target/linux/ath79/dts/qca9558_compex_wpj558-16m.dts:
```diff
@@ -0,0 +1,151 @@
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "compex,wpj558-16m", "qca,qca9558";
```
This is a dead link.
The 16m suffix should only be kept if there is an _official_ version with different flash size as well.

hamtinepeng replied:
The board was officially sold with a 8 MiB NOR (16 MiB was optional) however this version is not available on Compex website anymore.
I prefer keeping the 16 MiB reference because:
1) the device title in 19.07 specify the 16M version
2) the compex_wpj344-16m & compex_wpj531-16m board definition available in the same file follow the same principle
3) OpenWRT wiki talks about the old version

adschm replied:
> The board was officially sold with a 8 MiB NOR (16 MiB was optional) however this version is not available on Compex website anymore.

That's a valid argument.

> I prefer keeping the 16 MiB reference because:
> 
> 1. the device title in 19.07 specifies the 16M version

That's not.

> 2. the compex_wpj344-16m & compex_wpj531-16m board definition available in the same file follow the same principle

That's not.

> 3. OpenWRT wiki talks about the old version

Well, what does it tell.

Okay, jokes aside. Due to the one valid reason, you should keep the -16m suffix.

hamtinepeng commented:
New commit (099ddf7) solves previous issues/remarks

adschm commented:
Thanks, I will merge this later tonight.
